experimental/calendar/src/iso.rs

```diff
@@ -0,0 +1,310 @@
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Calendar, Date, DateDuration, DurationUnit};
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Copy, Clone, Debug)]
+/// The ISO Calendar
+pub struct Iso;
+
+/// A 1-indexed representation of an ISO day
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct IsoDay(u8);
+/// A 1-indexed representation of an ISO month
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct IsoMonth(u8);
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+/// An ISO year. Year 0 == 1 BCE
+pub struct IsoYear(pub i32);
+
+impl TryFrom<u8> for IsoDay {
+    type Error = ();
+    fn try_from(int: u8) -> Result<Self, ()> {
+        if int > 31 || int < 1 {
+            return Err(());
+        }
+        Ok(Self(int))
+    }
+}
+
+impl TryFrom<u8> for IsoMonth {
+    type Error = ();
+    fn try_from(int: u8) -> Result<Self, ()> {
+        if int > 12 || int < 1 {
+            return Err(());
+        }
+        Ok(Self(int))
+    }
+}
+
+impl From<i32> for IsoYear {
+    fn from(int: i32) -> Self {
+        Self(int)
+    }
+}
+
+impl From<IsoDay> for u8 {
+    fn from(day: IsoDay) -> Self {
+        day.0
+    }
+}
+
+impl From<IsoMonth> for u8 {
+    fn from(month: IsoMonth) -> Self {
+        month.0
+    }
+}
+
+impl From<IsoYear> for i32 {
+    fn from(year: IsoYear) -> Self {
+        year.0
+    }
+}
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub struct IsoDateInner {
+    day: IsoDay,
+    month: IsoMonth,
+    year: IsoYear,
+}
+
+impl IsoDateInner {
+    fn add_months(&mut self, months: i32) {
+        // Get a zero-indexed new month
+        let new_month = (self.month.0 as i32 - 1) + months;
+        if new_month >= 0 {
+            self.year.0 += new_month / 12;
+            self.month.0 = ((new_month % 12) + 1) as u8;
+        } else {
+            // subtract full years
+            self.year.0 -= (-new_month) / 12;
+            // subtract a partial year
+            self.year.0 -= 1;
+            // adding 13 since months are 1-indexed
+            self.month.0 = (13 + (new_month % 12)) as u8
+        }
+    }
+}
+
+impl Calendar for Iso {
+    type DateInner = IsoDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> IsoDateInner {
+        *iso.inner()
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        Date::construct_unchecked(*date, Iso)
+    }
+
+    fn months_in_year(&self, _date: &Self::DateInner) -> u8 {
+        12
+    }
+
+    fn days_in_year(&self, date: &Self::DateInner) -> u32 {
+        if Self::is_leap_year(date.year) {
+            366
+        } else {
+            365
+        }
+    }
+
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        Self::days_in_month(date.year, date.month)
+    }
+
+    fn day_of_week(&self, date: &Self::DateInner) -> u8 {
```

Comment:
Thought: We implement this function already at https://github.com/unicode-org/icu4x/blob/7ea67288c78dfe7ef56bfb702e2b6452bb13c17a/components/datetime/src/arithmetic.rs#L64.  Take a look and decide which implementation is better.

Reply:
Ditto about sharing/copying from icu_datetime in a followup.

Reply:
Nit: I would like if we recorded things like this more explicitly via TODO comments.  Could you open a follow-up issue for sharing more code between icu_datetime and icu_calendar and then reference it from here?

Reply:
I feel like the intent of this PR is to land an experimental partial implementation that I can iterate on, I don't really feel like I should file followups for all the things I plan to do to finish this. It'd make more sense IMO if this were not in experimental/ but part of the reason it's there is so that I can land partial stuff like this.

Reply:
But I can add a list of todos to the tracking issue